tools/codegen/generate.bzl

```diff
@@ -0,0 +1,120 @@
+load("@org_pubref_rules_protobuf//protobuf:rules.bzl", "proto_compile")
+load("@org_pubref_rules_protobuf//gogo:rules.bzl", "gogoslick_proto_library", "gogo_proto_library")
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
+
+def _impl(ctx):
+  m = []
+  for k, v in ctx.attr.importmap.items():
+    m += ["-m %s:%s" % (k, v)]
+
+  args = [ctx.file.src.path, "-o=" + ctx.outputs.out.path] + m
+  # print(args)
+
+  # Action to call the script.
+  ctx.action(
+      mnemonic="MixerGen",
+      inputs=[ctx.file.src],
+      outputs=[ctx.outputs.out],
+      arguments=args,
+      progress_message="Generating mixer go files in: %s" % ctx.outputs.out.path,
+      executable=ctx.executable._gen_tool)
+
+mixer_gen = rule(
+  implementation=_impl,
+  attrs={
+      "src": attr.label(allow_single_file=True),
+      "out": attr.output(mandatory=True),
+      "importmap": attr.string_dict(),
+      "_gen_tool": attr.label(executable=True, cfg="host", allow_files=True,
+                                default=Label("//tools/codegen/cmd/mixgenproc"))
+  },
+  # output_to_genfiles=True,
+)
+
+MIXER_DEPS = [
+    "//tools/codegen/pkg/adapter:go_default_library",
+    "//tools/codegen/pkg/template_extension:go_default_library",
+    "@com_github_istio_api//:mixer/v1/config/descriptor",
+]
+MIXER_INPUTS = [
+    "//tools/codegen/pkg/template_extension:proto",
+    "@com_github_istio_api//:mixer_descriptors",
+]
+MIXER_IMPORT_MAP = {
+    "mixer/v1/config/descriptor/value_type.proto": "istio.io/api/mixer/v1/config/descriptor",
+    "tools/codegen/pkg/template_extension/TemplateExtensions.proto": "istio.io/mixer/tools/codegen/pkg/template_extension",
+}
+MIXER_IMPORTS = [ "external/com_github_istio_api" ]
+
+# TODO: fill in with complete set of GOGO DEPS and IMPORT MAPPING
+GOGO_DEPS = [
+    "@com_github_gogo_protobuf//gogoproto:go_default_library",
+    "@com_github_gogo_protobuf//types:go_default_library",
+    "@com_github_gogo_protobuf//sortkeys:go_default_library",
+]
+GOGO_IMPORT_MAP = {
+    "gogoproto/gogo.proto": "github.com/gogo/protobuf/gogoproto",
+    "google/protobuf/duration.proto": "github.com/gogo/protobuf/types",
+}
+
+PROTO_IMPORTS = ["external/com_github_google_protobuf/src"]
+PROTO_INPUTS = [ "@com_github_google_protobuf//:well_known_protos" ]
+
+
+def mixer_proto_library(
+    name,
+    protos = [],
+    importmap = {},
```

Comment:
There probably is not one. I used those explicitly during development and then moved them all into generate.bzl default sets, as I figured they would be widely used.

If it is OK with you, we can add that sort of testing in a subsequent PR.